controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -0,0 +1,312 @@
+// deployment.
+func (w *Workload) UpdateCoreDNS(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane) error {
+	info, err := w.getCoreDNSInfo(ctx)
+	if err != nil {
+		return err
+	}
+	if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration == nil {
+		return nil
+	}
+
+	// Validate the image tag.
+	if err := validateCoreDNSImageTag(info.CurrentVersion, kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag); err != nil {
+		return errors.Wrapf(err, "failed to validate CoreDNS")
+	}
+
+	// Get the DNS cluster configuration.
+	dns := &kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS
+
+	// Perform the upgrade.
+	if err := w.updateCoreDNSImageInfoInKubeadmConfigMap(ctx, dns); err != nil {
+		return err
+	}
+	if err := w.updateCoreDNSCorefile(ctx, info, dns); err != nil {
+		return err
+	}
+	if err := w.updateCoreDNSDeployment(ctx, info, dns); err != nil {
+		return errors.Wrap(err, "unable to update coredns deployment")
+	}
+	return nil
+}
+
+// validateCoreDNSImageTag returns error if the versions don't meet requirements.
+// Some of the checks come from
+// https://github.com/coredns/corefile-migration/blob/v1.0.6/migration/migrate.go#L414
+func validateCoreDNSImageTag(fromVersion, toVersion string) error {
+	from, err := semver.Parse(fromVersion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to semver parse %q", fromVersion)
+	}
+	to, err := semver.Parse(toVersion)
+	if err != nil {
+		return errors.Wrapf(err, "failed to semver parse %q", toVersion)
+	}
+	if from.Compare(to) >= 0 {
+		return fmt.Errorf("toVersion %q must be greater than fromVersion %q", to.String(), from.String())
+	}
+
+	// check if the from version is even in the list of coredns versions
+	if _, ok := migration.Versions[fmt.Sprintf("%d.%d.%d", from.Major, from.Minor, from.Patch)]; !ok {
+		return fmt.Errorf("fromVersion %q is not a compatible coredns version", from.String())
+	}
+	return nil
+}
+
+// getCoreDNSInfo gets
+// - the current CoreDNS image version installed
+// - the current Corefile Configuration of CoreDNS
+// - the coredns deployment
+func (w *Workload) getCoreDNSInfo(ctx context.Context) (*coreDNSInfo, error) {
+	// Get the coredns configmap and corefile.
+	corednsKey := ctrlclient.ObjectKey{Name: coreDNSKey, Namespace: metav1.NamespaceSystem}
+	cm := &corev1.ConfigMap{}
+	if err := w.Client.Get(ctx, corednsKey, cm); err != nil {
+		return nil, errors.Wrapf(err, "error getting %s config map from target cluster", corednsKey.String())
+	}
+	corefile, ok := cm.Data[corefileKey]
+	if !ok {
+		return nil, errors.New("unable to find the CoreDNS Corefile data")
+	}
+
+	// Get the current CoreDNS deployment.
+	deployment := &appsv1.Deployment{}
+	if err := w.Client.Get(ctx, corednsKey, deployment); err != nil {
+		return nil, errors.Wrapf(err, "unable to get %s deployment from target cluster", corednsKey.String())
+	}
+
+	// Parse deployment image.
+	addonImage := deployment.Spec.Template.Spec.Containers[0].Image
```

Comment:
Seeing more and more hardcoding of `Containers[0]` is starting to worry me a bit, since there are no guarantees that there is only a single container defined or if multiple are defined the one we care about is the first.

I don't think it's a blocker for this PR, but we might want to create an issue to track trying to find a better long term solution to this.

Reply:
It also looks like we are already doing this in the Patch methods, just not currently when we are parsing the image from the deployment.

Reply:
Refactored this a bit, ptal